src/lib.rs

```diff
@@ -1884,6 +1884,10 @@ impl Connection {
         // Limit output packet size to respect peer's max_packet_size limit.
         left = cmp::min(left, max_pkt_len);
 
+        // Update app_limited.
+        self.recovery
+            .app_limited(left <= self.recovery.cwnd_available());
```

Comment:
The `left` value at this point simply represents the _maximum possible_ size of the output packet. It has nothing to do with how much data the application needs to send, so I don't understand how this represents the "app limited" state? I also don't think this takes flow control into consideration?

Reply:
At the very least this needs a much more detailed comment than "Update app_limited" as well as a test.

Reply:
Good point. The reason why I put it here is to update app_limited even send() returns Error::Done. It covers most of the case but there are some exceptions. Let me think a little more.

Reply:
Updated and also added test cases.

Reply:
Recent recovery draft code need to check app limited and flow limited both, however this PR is to fix app limited. For flow control it's better to do in a separate PR.

Reply:
So, if I understand the new changes correctly the idea is to update `app_limited` when we exit early without sending a packet?

I think in both of the cases covered the only reason why we wouldn't send a packet is a too small cwnd, so the `hdr_len` checks seem redundant and we might just want to set `app_limited` to `false` in both cases?

Reply:
I changed first case to simply set app_limited=false. However second case (when no frame is generated) it still need to check with cwnd_available(). (e.g. when we have nothing to send but cwnd has a room, app_limited is true). Also added one more test case.

Reply:
If there is nothing to write we will return much earlier, when `write_epoch()` is called, so we wouldn't reach this point, and I think the check always ends up evaluating to `false`.

Reply:
Thanks. Fixed!